keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -378,6 +384,200 @@ class SqlArtifactRepository(
+      .mapNotNull { artifact ->
+        when (val uid = artifact.uidString) {
+          null -> null
+          else -> uid to artifact
+        }
+      }
+      .toMap()
+
+    val vetoes: MutableMap<String, EnvironmentArtifactVetoes> = mutableMapOf()
+
+    jooq.select(
+      ENVIRONMENT.NAME,
+      ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID,
+      ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION
+    )
+      .from(ENVIRONMENT)
+      .innerJoin(ENVIRONMENT_ARTIFACT_VERSIONS)
+      .on(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+      .where(
+        ENVIRONMENT.DELIVERY_CONFIG_UID.eq(deliveryConfig.uid),
+        ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.eq(VETOED.name)
+      )
+      .fetch { (envName, artifactId, version) ->
+        if (artifactsById.containsKey(artifactId)) {
+          vetoes.getOrPut(vetoKey(envName, artifactId),
+            {
+              EnvironmentArtifactVetoes(
+                deliveryConfigName = deliveryConfig.name,
+                targetEnvironment = envName,
+                artifact = artifactsById[artifactId]
+                  ?: error("Invalid artifactId $artifactId for deliveryConfig ${deliveryConfig.name}"),
+                versions = mutableSetOf())
+            })
+            .versions.add(version)
+        }
+      }
+
+    return vetoes.values.toList()
+  }
+
+  override fun markAsVetoedIn(
+    deliveryConfig: DeliveryConfig,
+    artifact: DeliveryArtifact,
+    version: String,
+    targetEnvironment: String,
+    force: Boolean
+  ): Boolean {
+
+    val (envUid, artUid) = sqlRetry.withRetry(READ) {
+      Pair(
+        deliveryConfig.getUidStringFor(
+          deliveryConfig.environmentNamed(targetEnvironment)),
+        artifact.uidString)
+    }
+
+    val isPinned = sqlRetry.withRetry(READ) {
+      jooq
+        .fetchExists(
+          ENVIRONMENT_ARTIFACT_PIN,
+          ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID.eq(envUid)
+            .and(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID.eq(artUid))
+            .and(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION.eq(version)))
+    }
+
+    if (isPinned) {
+      log.warn(
+        "Pinned artifact version cannot be vetoed: " +
+          "deliveryConfig=${deliveryConfig.name}, " +
+          "environment=$targetEnvironment, " +
+          "artifactVersion=$version")
+      return false
+    }
+
+    return jooq
+      .select(
+        ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS,
+        ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_REFERENCE)
+      .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+      .where(
+        ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(envUid),
+        ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artUid),
+        ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(version))
+      .fetchOne { (status, reference) ->
+        if (reference != null && !force) {
+          log.warn(
+            "Not vetoing artifact version as it appears to have already been an automated rollback target: " +
```

Comment:
Is this true regardless of the value of `status`?

Reply:
Yep.